## Housing Price Prediction

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### Reading the Dataset and Data Exploration

In [12]:
df = pd.read_csv("melb_data.csv")
df.head()

Unnamed: 0      Suburb           Address  Rooms Type      Price Method  \
0           1  Abbotsford      85 Turner St      2    h  1480000.0      S   
1           2  Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2           4  Abbotsford      5 Charles St      3    h  1465000.0     SP   
3           5  Abbotsford  40 Federation La      3    h   850000.0     PI   
4           6  Abbotsford       55a Park St      4    h  1600000.0     VB   

  SellerG       Date  Distance  ...  Bathroom  Car  Landsize  BuildingArea  \
0  Biggin  3/12/2016       2.5  ...       1.0  1.0     202.0           NaN   
1  Biggin  4/02/2016       2.5  ...       1.0  0.0     156.0          79.0   
2  Biggin  4/03/2017       2.5  ...       2.0  0.0     134.0         150.0   
3  Biggin  4/03/2017       2.5  ...       2.0  1.0      94.0           NaN   
4  Nelson  4/06/2016       2.5  ...       1.0  2.0     120.0         142.0   

   YearBuilt  CouncilArea  Lattitude Longtitude             Regionname  \
0        NaN        Yarra   -37.7996   144.9984  Northern Metropolitan   
1     1900.0        Yarra   -37.8079   144.9934  Northern Metropolitan   
2     1900.0        Yarra   -37.8093   144.9944  Northern Metropolitan   
3        NaN        Yarra   -37.7969   144.9969  Northern Metropolitan   
4     2014.0        Yarra   -37.8072   144.9941  Northern Metropolitan   

   Propertycount  
0         4019.0  
1         4019.0  
2         4019.0  
3         4019.0  
4         4019.0  

[5 rows x 22 columns]

In [15]:
nan = df.isnull().sum()
nan

Unnamed: 0           0
Suburb               0
Address              0
Rooms                0
Type                 0
Price                0
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          3469
Bathroom          3471
Car               3576
Landsize          4793
BuildingArea     10634
YearBuilt         9438
CouncilArea       6163
Lattitude         3332
Longtitude        3332
Regionname           1
Propertycount        1
dtype: int64

In [16]:
df.describe()

Unnamed: 0         Rooms         Price      Distance      Postcode  \
count  18396.000000  18396.000000  1.839600e+04  18395.000000  18395.000000   
mean   11826.787073      2.935040  1.056697e+06     10.389986   3107.140147   
std     6800.710448      0.958202  6.419217e+05      6.009050     95.000995   
min        1.000000      1.000000  8.500000e+04      0.000000   3000.000000   
25%     5936.750000      2.000000  6.330000e+05      6.300000   3046.000000   
50%    11820.500000      3.000000  8.800000e+05      9.700000   3085.000000   
75%    17734.250000      3.000000  1.302000e+06     13.300000   3149.000000   
max    23546.000000     12.000000  9.000000e+06     48.100000   3978.000000   

           Bedroom2      Bathroom           Car       Landsize  BuildingArea  \
count  14927.000000  14925.000000  14820.000000   13603.000000   7762.000000   
mean       2.913043      1.538492      1.615520     558.116371    151.220219   
std        0.964641      0.689311      0.955916    3987.326586    519.188596   
min        0.000000      0.000000      0.000000       0.000000      0.000000   
25%        2.000000      1.000000      1.000000     176.500000     93.000000   
50%        3.000000      1.000000      2.000000     440.000000    126.000000   
75%        3.000000      2.000000      2.000000     651.000000    174.000000   
max       20.000000      8.000000     10.000000  433014.000000  44515.000000   

         YearBuilt     Lattitude    Longtitude  Propertycount  
count  8958.000000  15064.000000  15064.000000   18395.000000  
mean   1965.879996    -37.809849    144.996338    7517.975265  
std      37.013261      0.081152      0.106375    4488.416599  
min    1196.000000    -38.182550    144.431810     249.000000  
25%    1950.000000    -37.858100    144.931193    4294.000000  
50%    1970.000000    -37.803625    145.000920    6567.000000  
75%    2000.000000    -37.756270    145.060000   10331.000000  
max    2018.000000    -37.408530    145.526350   21650.000000

In [17]:
## Treating NaN
df = df.dropna(axis=0)
df

Unnamed: 0        Suburb          Address  Rooms Type      Price  \
1               2    Abbotsford  25 Bloomburg St      2    h  1035000.0   
2               4    Abbotsford     5 Charles St      3    h  1465000.0   
4               6    Abbotsford      55a Park St      4    h  1600000.0   
6              11    Abbotsford     124 Yarra St      3    h  1876000.0   
7              14    Abbotsford    98 Charles St      2    h  1636000.0   
...           ...           ...              ...    ...  ...        ...   
15388       19732    Whittlesea    30 Sherwin St      3    h   601000.0   
15389       19733  Williamstown      75 Cecil St      3    h  1050000.0   
15390       19734  Williamstown    2/29 Dover Rd      1    u   385000.0   
15392       19736       Windsor  201/152 Peel St      2    u   560000.0   
15395       19739    Yarraville  54 Pentland Pde      6    h  2450000.0   

      Method        SellerG        Date  Distance  ...  Bathroom  Car  \
1          S         Biggin   4/02/2016       2.5  ...       1.0  0.0   
2         SP         Biggin   4/03/2017       2.5  ...       2.0  0.0   
4         VB         Nelson   4/06/2016       2.5  ...       1.0  2.0   
6          S         Nelson   7/05/2016       2.5  ...       2.0  0.0   
7          S         Nelson   8/10/2016       2.5  ...       1.0  2.0   
...      ...            ...         ...       ...  ...       ...  ...   
15388      S            Ray  29/07/2017      35.5  ...       2.0  1.0   
15389     VB       Williams  29/07/2017       6.8  ...       1.0  0.0   
15390     SP       Williams  29/07/2017       6.8  ...       1.0  1.0   
15392     PI  hockingstuart  29/07/2017       4.6  ...       1.0  1.0   
15395     VB        Village  29/07/2017       6.3  ...       3.0  2.0   

       Landsize  BuildingArea  YearBuilt  CouncilArea  Lattitude Longtitude  \
1         156.0         79.00     1900.0        Yarra  -37.80790  144.99340   
2         134.0        150.00     1900.0        Yarra  -37.80930  144.99440   
4         120.0        142.00     2014.0        Yarra  -37.80720  144.99410   
6         245.0        210.00     1910.0        Yarra  -37.80240  144.99930   
7         256.0        107.00     1890.0        Yarra  -37.80600  144.99540   
...         ...           ...        ...          ...        ...        ...   
15388     972.0        149.00     1996.0   Whittlesea  -37.51232  145.13282   
15389     179.0        115.00     1890.0  Hobsons Bay  -37.86558  144.90474   
15390       0.0         35.64     1967.0  Hobsons Bay  -37.85588  144.89936   
15392       0.0         61.60     2012.0  Stonnington  -37.85581  144.99025   
15395    1087.0        388.50     1920.0  Maribyrnong  -37.81038  144.89389   

                  Regionname  Propertycount  
1      Northern Metropolitan         4019.0  
2      Northern Metropolitan         4019.0  
4      Northern Metropolitan         4019.0  
6      Northern Metropolitan         4019.0  
7      Northern Metropolitan         4019.0  
...                      ...            ...  
15388      Northern Victoria         2170.0  
15389   Western Metropolitan         6380.0  
15390   Western Metropolitan         6380.0  
15392  Southern Metropolitan         4380.0  
15395   Western Metropolitan         6543.0  

[6196 rows x 22 columns]

### Machine Learning

In [19]:
## Defyning Variables
# y to predict
y = df['Price']
# X independent
X = df[['Rooms', 'Bathroom', 'Landsize', 'Lattitude', 'Longtitude']]

In [20]:
from sklearn.tree import DecisionTreeRegressor
## Creating the model
melb_model = DecisionTreeRegressor()
## Train the model
melb_model.fit(X, y)

DecisionTreeRegressor()

In [22]:
## Printing out
print("Prediction of The First Five House")
print(X.head())
print("The Real Price is:")
print(y.head())
print("Predictions Are:")
print(melb_model.predict(X.head()))

Prediction of The First Five House
   Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2       1.0     156.0   -37.8079    144.9934
2      3       2.0     134.0   -37.8093    144.9944
4      4       1.0     120.0   -37.8072    144.9941
6      3       2.0     245.0   -37.8024    144.9993
7      2       1.0     256.0   -37.8060    144.9954
The Real Price is:
1    1035000.0
2    1465000.0
4    1600000.0
6    1876000.0
7    1636000.0
Name: Price, dtype: float64
Predictions Are:
[1035000. 1465000. 1600000. 1876000. 1636000.]


### MAE (MEAN ABSOLUTE ERROR)

VERIFICATION (ERRORS)

In [23]:
from sklearn.metrics import mean_absolute_error

predicted_home_prices = melb_model.predict(X)
mean_absolute_error(y, predicted_home_prices)

1115.7467183128902

### VALIDATION


In [27]:
## To evaluate the overfitting
from sklearn.model_selection import train_test_split

train_X, val_X, train_y, val_y = train_test_split(X, y)

melb_model = DecisionTreeRegressor()

melb_model.fit(train_X, train_y)

val_predictions = melb_model.predict(val_X)
print(mean_absolute_error(val_y, val_predictions))

244907.88250484184


Model failed with a price of 244,907 EUR

### Fuctions to fix the error

By creation a function where the model is the decisionTreeRegressor, but I need to change the max_leaf_node, predict X's values, and going to return MAE. This process is just to prove different shet's numbers in order to get better results

In [28]:
def get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y):
    model = DecisionTreeRegressor(max_leaf_nodes=max_leaf_nodes)
    model.fit(train_X, train_y)
    preds_val = model.predict(val_X)
    mae = mean_absolute_error(val_y, preds_val)
    return(mae)

In [30]:
for max_leaf_nodes in [5, 50, 500, 5000]:
    my_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y)
    print("Max leaf nodes: %d  \t\t Mean Absolute Error:  %d" %(max_leaf_nodes, my_mae))

Max leaf nodes: 5  		 Mean Absolute Error:  373690
Max leaf nodes: 50  		 Mean Absolute Error:  259584
Max leaf nodes: 500  		 Mean Absolute Error:  236680
Max leaf nodes: 5000  		 Mean Absolute Error:  244768


#### Documentation:

* With a leaf nodes of 5 sheet, we get an error of: 373,690
* With a leaf nodes of 50 sheet, we get an error of:259,584
* With a leaf nodes of 500 sheet, we get an error of:236,680 --> This one being the better result in DecisionTree Model
* With a leaf nodes of 5000 sheet, we get an error of:244,768

### Random Forest Regressor

In [34]:
import numpy as np
## Reshape on val_y for random forest which only accept 2D
val_y_reshaped = val_y.to_numpy().reshape(-1, 1)

In [37]:
from sklearn.ensemble import RandomForestRegressor

forest_model = RandomForestRegressor()
forest_model.fit(train_X, train_y)
melb_preds = forest_model.predict(val_X)
print(mean_absolute_error(val_y, melb_preds))

182755.79784530727


### Documentation

Based on Random Forest Regressor,
The result of the mean absolute error is 182,755 EUR, which mean is 53,925 less becoming it better than DecisionTreeRegressor Model.